### ISSUE TITLE
Can't set the default value for multiple options with optional values

### ISSUE THREAD

**Issue by ikapelyukhin** [pallets/click#2001]
The following example combines [optional values](https://click.palletsprojects.com/en/8.0.x/options/#optional-value) with [multiple options](https://click.palletsprojects.com/en/8.0.x/options/#multiple-options):

```python
#!/usr/bin/env python3

import click


@click.command()
@click.option('--label', multiple=True, is_flag=False, flag_value='default')
def cli(label):
    print(label)
    print(f"click.core._flag_needs_value: {click.core._flag_needs_value}")


cli()
```

```
$ ./test.py --label --label test
('<object object at 0x7f5ef63ec250>', 'test')
click.core._flag_needs_value: <object object at 0x7f5ef63ec250>
```

Expected result: `('default', 'test')`

Looks like when `multiple=True`, the sentinel value (`click.core._flag_needs_value`) is never substituted with the default value specified by `flag_value`.

Environment:

- Python version: 3.7.3
- Click version: 8.0.1

**Comment by janluke** [comment]
Is this thing of using an option both as a flag and a non-flag common practice? I've never seen something like that in the CLIs I've used. Personally I wouldn't even allow it.

**Comment by davidism** [comment]
Yeah, this doesn't seem like a feature we were intending to support. I'll probably close this, if someone wants to make a PR that makes it work consistently I'm open to it, but I don't think it needs to be an open issue.

**Comment by ikapelyukhin** [comment]
> Is this thing of using an option both as a flag and a non-flag common practice?

It's probably rather uncommon, but it is something I've run into. And I'm not using it as a flag per se: in my (somewhat strange, I admit) use case it is possible to create a label without any text, so `--label` is basically a shorthand for `--label ""`. This also happens to work perfectly fine when `multiple=False`.

> Yeah, this doesn't seem like a feature we were intending to support. 

Well, if it isn't supported then it would be nice to get an exception when conflicting options (`is_flag=False` and `multiple=True`) are specified. Otherwise you get `'<object object at 0x7f5ef63ec250>'` as the option value instead of the default, that doesn't seem like correct behavior.

> I'll probably close this, if someone wants to make a PR that makes it work consistently I'm open to it, but I don't think it needs to be an open issue.

I can undo all changes related to prompting in #2004 for options with multiple values, then `ParameterSource` would be consistent.

**Comment by davidism** [comment]
We don't generally validate every single possible combination of things. There are a *ton* of combinations, and as a CLI application it has to be reasonably fast to create everything so it can execute when the user types the command.

I'm not clear why you want to roll back #2004, but you're welcome to do that if you want, just keep in mind our license.

**Comment by ikapelyukhin** [comment]
> I'm not clear why you want to roll back #2004.

I don't mean the whole PR, I've made some changes that made things weird when it came to keeping `ParameterSource` consistent.

**Comment by ikapelyukhin** [comment]
OK, I've made the changes that I wanted, so #2004 is good to go from my perspective.
